source/module_hsolver/hsolver_pw.cpp

```diff
@@ -81,19 +83,40 @@ void HSolverPW<T, Device>::initDiagh(const psi::Psi<T, Device>& psi)
     }
     else if (this->method == "dav")
     {
-        DiagoDavid<T>::PW_DIAG_NDIM = GlobalV::PW_DIAG_NDIM;
         if (this->pdiagh != nullptr)
         {
             if (this->pdiagh->method != this->method)
             {
                 delete (DiagoDavid<T, Device>*)this->pdiagh;
-                this->pdiagh = new DiagoDavid<T, Device>(precondition.data());
+
+
+                this->pdiagh = new DiagoDavid<T, Device>(
+                                precondition.data(),
+                                GlobalV::PW_DIAG_NDIM,
+                                GlobalV::use_paw,
+#ifdef __MPI                     
+                                {POOL_WORLD, GlobalV::RANK_IN_POOL, GlobalV::NPROC_IN_POOL}
+#else
+                                {}
+#endif
+                                );
+
                 this->pdiagh->method = this->method;
             }
         }
         else
         {
-            this->pdiagh = new DiagoDavid<T, Device>(precondition.data());
+            this->pdiagh = new DiagoDavid<T, Device>(
+                                precondition.data(),
+                                GlobalV::PW_DIAG_NDIM,
+                                GlobalV::use_paw,
+#ifdef __MPI                     
```

Comment:
can you initialize an empty object of "diag_comm_info"? then you can avoid using "#ifdef __MPI"